### PR TITLE
Redesign the architecture of fmn.consumer

### PR DESCRIPTION
**Pull request by pypingou**

### CODE REVIEW

**Comment by skrzepto** [comment]
Just rebased `sse` branch with `split_process` now and all is running well :)

File: README.rst [hunk 1]
```diff
@@ -18,10 +18,146 @@ HACKING
...
+* the backend
+
+The consumer is the part of the application that receives the messages from
+fedmsg, it then sends them to the worker(s) who will see whose preferences
+the message match (ie: to whom to send the notification and on which channel).
+The worker(s) then send its results to the backend that receives them and
+just do the IO: sending the email, posting to IRC and so on.
+
+
+To run these parts, simply call, in three different terminals:
```
Comment (puiterwijk) [inline README.rst]: Perhaps add a short note here that people should run rabbitmq?
I see it's further down in the FAQ as well, but just a quick note here might be useful.

File: README.rst [hunk 2]
```diff
@@ -18,10 +18,146 @@ HACKING
...
+
+::
+
+    $ fedmsg-hub
```
Comment (puiterwijk) [inline README.rst]: Probably need to also add that this should all be done in the virtualenv?

File: fmn/consumer/backend.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+log = logging.getLogger("fmn")
+log.setLevel('DEBUG')
+CONFIG = fedmsg.config.load_config()
+fedmsg.meta.make_processors(**CONFIG)
+
+DB_URI = CONFIG.get('fmn.sqlalchemy.uri', None)
+session = fmn.lib.models.init(DB_URI)
+
+fmn.consumer.fmn_fasshim.make_fas_cache(**CONFIG)
+# Duck patch fedmsg_meta modules
```
Comment (puiterwijk) [inline fmn/consumer/backend.py]: I have never heard of duck-patching. Sure you don't mean monkey-patching?

File: fmn/consumer/backend.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+parameters = pika.ConnectionParameters()
+cc = protocol.ClientCreator(
+    reactor, twisted_connection.TwistedProtocolConnection, parameters)
+d = cc.connectTCP('localhost', 5672)
+d.addCallback(lambda protocol: protocol.ready)
+d.addCallback(run)
+
+# Here we schedule to producers to run periodically (with a default
+# frequency of 10 seconds.
+# Added value: Everything is nicely tight up with twisted in one app/place
```
Comment (puiterwijk) [inline fmn/consumer/backend.py]: s/tight/tied/?

File: fmn/consumer/consumer.py [hunk 5]
```diff
@@ -18,6 +18,39 @@
...
 
+import pika
+connection = pika.BlockingConnection()
+
+
+def notify_prefs_change(openid):
+    import json
+    import pika
+    connection = pika.BlockingConnection()
+    msg_id = '2016-%s' % uuid.uuid4()
```
Comment (puiterwijk) [inline fmn/consumer/consumer.py]: I guess the year should be a variable that comes from a datetime.date.today().year?

File: fmn/consumer/fmn_fasshim.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            _cache.set(nick, user['username'])
+
+        email = user['email']
+        if email:
+            _cache.set(email, user['username'])
+    else:
+        # If we couldn't find the nick in FAS, save it in the _cache as nick
+        # so that we avoid calling FAS for every single filter we have to
+        # run through
+        _cache.set(username, username)
```
Comment (puiterwijk) [inline fmn/consumer/fmn_fasshim.py]: Perhaps set it explicitly to None?

File: fmn/consumer/backend.py (hunk 3 above)
Comment (pypingou) [inline fmn/consumer/backend.py]: You know if it quacks like a duck, it's a duck :-D

File: fmn/consumer/consumer.py (hunk 5 above)
Comment (pypingou) [inline fmn/consumer/consumer.py]: likely yes

File: fmn/consumer/backend.py (hunk 4 above)
Comment (pypingou) [inline fmn/consumer/backend.py]: ok

File: README.rst (hunk 2 above)
Comment (pypingou) [inline README.rst]: line 37 then, no?

File: README.rst (hunk 1 above)
Comment (pypingou) [inline README.rst]: sure

File: fmn/consumer/backend.py (hunk 3 above)
Comment (puiterwijk) [inline fmn/consumer/backend.py]: Almost, you can't be sure! :)

File: README.rst (hunk 2 above)
Comment (puiterwijk) [inline README.rst]: Correct

File: fmn/consumer/worker.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    fmn.consumer.fmn_fasshim.email2fas
+fedmsg_meta_fedora_infrastructure.mailman3.email2fas = \
+    fmn.consumer.fmn_fasshim.email2fas
+fedmsg_meta_fedora_infrastructure.pagure.email2fas = \
+    fmn.consumer.fmn_fasshim.email2fas
+
+connection = pika.BlockingConnection(OPTS)
+
+def inform_workers(raw_msg, context, recipients):
+    queue = 'backends'
```
Comment (puiterwijk) [inline fmn/consumer/worker.py]: Perhaps in the future a backend per context.
But that's for the future I'd say.

Comment (pypingou) [inline fmn/consumer/worker.py]: Yes, for now I want one queue and I want to monitor it, based on this info we'll split or not

**Comment by puiterwijk** [comment]
:shipit:

**Comment by pypingou** [comment]
Cool, thanks!! :)

**Pull request merged**